### ISSUE TITLE
cmd/unit.cpp: Pointer cast to `long` instead of to `intptr_t` (5 occurrences)

### ISSUE THREAD

**Issue by stephengtuggy** [vegastrike/Vega-Strike-Engine-Source#522]
In `GameUnit::Draw(const Transformation &, const Matrix &)`, in the section toward the bottom of the function that draws particles based on level of damage, there are 5 lines that cast pointer `this` to `long` to use as a random seed:

```C++
                switch (switcher)
                {
                case 5:
                    LaunchOneParticle( *ctm, this->GetVelocity(), ( (long) this )+165, this, damagelevel, this->faction );
                case 4:
                    LaunchOneParticle( *ctm, this->GetVelocity(), ( (long) this )+47, this, damagelevel, this->faction );
                case 3:
                    LaunchOneParticle( *ctm, this->GetVelocity(), ( (long) this )+61, this, damagelevel, this->faction );
                case 2:
                    LaunchOneParticle( *ctm, this->GetVelocity(), ( (long) this )+65537, this, damagelevel, this->faction );
                default:
                    LaunchOneParticle( *ctm, this->GetVelocity(), ( (long) this )+257, this, damagelevel, this->faction );
                }
```

The Visual C++ compiler warns about these 5 lines that there may be a loss of precision. I believe this is because the `this` pointer should be cast to `intptr_t`, not to `long`. (Because `long` is signed, whereas x86_64 pointers are not?)

**Comment by BenjamenMeyer** [comment]
The cast to `long` should be removed; we should be casting to a bit type instead.

**Comment by stephengtuggy** [comment]
> The cast to `long` should be removed; we should be casting to a bit type instead.

@BenjamenMeyer Pardon my ignorance; what do you mean by a bit type?

**Comment by BenjamenMeyer** [comment]
@stephengtuggy (u)int(8/16/32/64) types

**Comment by stephengtuggy** [comment]
@BenjamenMeyer Ah, OK. So you're talking about the types listed [here](https://en.cppreference.com/w/cpp/types/integer) .

The correct choice from that list, I believe, is `uintptr_t`. If it's present in a given C++ implementation, then it represents an unsigned integer type large enough to hold a pointer.

**Comment by BenjamenMeyer** [comment]
@stephengtuggy yes; if it's for holding a pointer then yes `uintptr_t` would be correct; if not, then one of the other types. Either way it also helps document the usage too.

**Comment by stephengtuggy** [comment]
I suppose `uintmax_t` would also work.

**Comment by BenjamenMeyer** [comment]
> I suppose `uintmax_t` would also work.

depends on the use. granted most systems are 64-bit so that would be equivalent to uint64_t; however, if we're concerned about data ranges then it needs to be a specific bit-size since uintmax_t would be 32-bit on a 32-bit system - RPi2/3, Intel Atom, etc...yes, there's still a lot of cheap Intel/x86 procs that are 32-bit only, though hopefully those are going away as far as new sales goes. Older systems, which will probably favor Vega Strike more, will more likely see a large group of 32-bit systems.

So unless there's a specific reason not to, I'd prefer to use a fixed bit integer type so that it's well known and the range can be relied upon.

**Comment by stephengtuggy** [comment]
What we are doing here, ultimately, is we are taking a pointer value; treating it as an `unsigned int`; and adding an integer literal to it. We then pass the resulting `unsigned int` to the `LaunchOneParticle` function, to be used as a random seed.

The pointer value will be different sizes on different systems. `unsigned int` will also be different sizes on different systems. I don't want to change `LaunchOneParticle`'s signature yet, in case it's used directly by the game assets.

Initially I was thinking we would perform two casts, one to `uintptr_t`, and then a second to `unsigned int`. Does that make sense? Or could we perhaps just do this?

```C++
static_cast<unsigned int>(this) + 47
```

etc.?

**Comment by stephengtuggy** [comment]
Well, I tried a C++11-style `static_cast` to `uintptr_t`. That didn't work. Maybe I need to try a different style of cast or something.
